### ISSUE TITLE
lnms smokeping:generate exits on error instead of skipping misconfigured host breaking smokeping.

### ISSUE THREAD

**Issue by IKatsu** [librenms/librenms#12778]
I have a multi-user librenms install and someone had to stop while adding more hosts to it and left one misconfigured. (empty hostname in this case)
This broke the smokeping config generator script, instead of skipping a bad host (which i think is the intent when reading the code) it exits with an error because that's how lnms works.
I can't push a commit and I doubt I'll ever have to again so here's a quick fix patch file.
Original error was:
In SmokepingGenerateCommand.php line 348:
checkdnsrr(): Host cannot be empty

```
diff --git a/app/Console/Commands/SmokepingGenerateCommand.php b/app/Console/Commands/SmokepingGenerateCommand.php
index cc0809a03..fd57c9a13 100644
--- a/app/Console/Commands/SmokepingGenerateCommand.php
+++ b/app/Console/Commands/SmokepingGenerateCommand.php
@@ -344,7 +344,11 @@ class SmokepingGenerateCommand extends LnmsCommand
     private function deviceIsResolvable($hostname)
     {
         // First we check for IP literals, then for a dns entry, finally for a hosts entry due to a PHP/libc limitation
-        // We look for the hosts entry last (and separately) as this only works for v4 - v6 host entries won't be found
+       // We look for the hosts entry last (and separately) as this only works for v4 - v6 host entries won't be found
+       if (empty($hostname)) { 
+           //But lets not completely break config configuration over 1 misconfigured host and instead just skip it.
+           return false;
+       }
         if (filter_var($hostname, FILTER_VALIDATE_IP) || checkdnsrr($hostname, 'ANY') || is_array(gethostbynamel($hostname))) {
             return true;
         }
```

**Comment by rhinoau** [comment]
> **TL;DR**
> 1. Do you have the no-dns option set?
> 2. Can you test the diff at the bottom and does it work instead?


Early days with me and php and didn't feel confident doing a PR of this until I understood it better, I think I've got near the bottom of the rabbit hole now! ;)

As that function currently stands, an empty string fails the test as it should:

```
php > $hostname='';

php > var_dump(empty($hostname));
bool(true)

php > var_dump(filter_var($hostname, FILTER_VALIDATE_IP) || checkdnsrr($hostname, 'ANY') || is_array(gethostbynamel($hostname)));
PHP Warning:  checkdnsrr(): Host cannot be empty in php shell code on line 1
bool(false)
```

This function is called here on line 320 and would fail, so the ->dnsLookup call must be coming back false:

https://github.com/librenms/librenms/blob/7c157c1577c7d7dceb01269114f100d08609b911/app/Console/Commands/SmokepingGenerateCommand.php#L315-L332

Proof:
```
php > var_dump(! false || false );
bool(true)
php > var_dump(! true || false);
bool(false)
```

I'm not across PHP object/inheritance syntax yet, but is this hard-coding a default?  Meaning it must be overridden by the no-dns option in order for the above to be true?

https://github.com/librenms/librenms/blob/7c157c1577c7d7dceb01269114f100d08609b911/app/Console/Commands/SmokepingGenerateCommand.php#L32-L35
https://github.com/librenms/librenms/blob/7c157c1577c7d7dceb01269114f100d08609b911/app/Console/Commands/SmokepingGenerateCommand.php#L285-L287
https://github.com/librenms/librenms/blob/7c157c1577c7d7dceb01269114f100d08609b911/app/Console/Commands/SmokepingGenerateCommand.php#L101-L104

Does this then mean that you've got no-dns set, and it is only in this scenario that the issue arises? If so, then the correct place to fix this would maybe be the line 320 call with something which should be far better written than this: 

`(    (! $this->dnsLookup || $this->deviceIsResolvable($hostname)    )  &&  ! empty($hostname)    )`

If you apply this against master - does it solve your issue?

```
diff --git a/app/Console/Commands/SmokepingGenerateCommand.php b/app/Console/Commands/SmokepingGenerateCommand.php
index cc0809a03..644da9f67 100644
--- a/app/Console/Commands/SmokepingGenerateCommand.php
+++ b/app/Console/Commands/SmokepingGenerateCommand.php
@@ -317,7 +317,7 @@ class SmokepingGenerateCommand extends LnmsCommand
         $lines = [];

         foreach ($devices as $hostname => $config) {
-            if (! $this->dnsLookup || $this->deviceIsResolvable($hostname)) {
+            if ( (! $this->dnsLookup || $this->deviceIsResolvable($hostname)) && !empty($hostname)) {
                 $lines[] = sprintf('++ %s', $this->buildMenuEntry($hostname));
                 $lines[] = sprintf('   menu = %s', $hostname);
                 $lines[] = sprintf('   title = %s', $hostname);
```

**Comment by murrant** [comment]
@rhinoau Yes, correct fix would be to add a check for empty hostnames before trying to lookup a dns record.

I would probably even put it inside the deviceIsResolvable() function (because we know '' is not resolvable)
